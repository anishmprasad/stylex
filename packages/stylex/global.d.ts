// @author: somibuon, Duc Hoang
// TODO: need more review

import * as css from 'csstype';

export type CssValue =
  | string
  | number
  | null
  | false

type NormalCssProperties = css.Properties<string | number>
type NormalCssValues<K> = K extends keyof NormalCssProperties
  ? NormalCssProperties[K] | CssValue
  : CssValue
type NormalCssStyle = {
  [K in keyof NormalCssProperties]: NormalCssValues<K> | CssStyle
}

/**
 * This interface can be augmented by users to add default types for custom css properties when
 * using `@ladifire-opensource/stylex/global`.
 * Use module augmentation to append your own type definition in a your_custom_type.d.ts file.
 * https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation
 */
interface CustomCssProperties extends NormalCssStyle {
}

export type CssStyle =
  | NormalCssStyle
  | CustomCssProperties
  | {
  [K: string]: NormalCssStyle | CssStyle
}

export type Style<Name extends string | number | symbol = string> = Record<
  Name,
  | CssStyle
  | string
  >
