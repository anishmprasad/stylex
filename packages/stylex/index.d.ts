// @author: somibuon, Duc Hoang
// TODO: need more review

import {CssStyle, Style} from './global';

// tslint:disable-next-line:export-just-namespace
export = stylex;
export as namespace stylex;

declare function stylex(...style: (CssStyle | CssStyle[])[]): string;

declare namespace stylex {
  function create<T extends string | number>(styles: Style<T>) : { [key in T]: CssStyle }

  function dedupe(...styles: CssStyle[]): string;

  // TODO: maybe remove in next release
  function compose(...styles: CssStyle[]): CssStyle;

  function keyframes(rules: CssStyle): string;

  function inject(cssString: string): void;
}
