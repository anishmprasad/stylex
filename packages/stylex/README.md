# @ladifire-opensource/stylex

This package is part of [stylex](https://github.com/ladifire-opensource/stylex). See the repo for more details.