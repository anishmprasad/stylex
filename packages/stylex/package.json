{
  "name": "@ladifire-opensource/stylex",
  "version": "0.0.1-beta.3.0",
  "description": "Use CSS in JS with atomic support",
  "license": "MIT",
  "publishConfig": {
    "access": "public"
  },
  "main": "dist/main.js",
  "module": "dist/main.module.js",
  "types": "index.d.ts",
  "source": "index.js",
  "files": [
    "dist",
    "src",
    "index.d.ts",
    "global.d.ts",
    "index.js",
    "stylex.js"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/ladifire-opensource/stylex"
  },
  "dependencies": {
    "@ladifire-opensource/stylex-theme": "0.0.1-beta.3.0"
  }
}
