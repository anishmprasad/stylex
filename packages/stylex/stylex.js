/**
 * Copyright (c) Ladifire, Inc. And its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

const CometStyleXSheet = require('@ladifire-opensource/stylex-theme');

CometStyleXSheet.rootStyleSheet.injectTheme();

var g = false;

function h(a) {
  a = a.reverse();
  const b = {};
  while (a.length) {
    var c = a.pop();
    if (Array.isArray(c)) {
      for (var d = c.length - 1; d >= 0; d--) {a.push(c[d]);}
      continue;
    }
    d = c;
    if (d != null && typeof d === 'object') {
      for (var e in d) {
        c = d[e];
        if (typeof c === 'string') {b[e] = c;} else if (typeof c === 'object') {
          var f;
          b[e] = (f = b[e]) != null ? f : {};
          Object.assign(b[e], c);
        }
      }
    }
  }
  return b;
}

function stylex(...args) {
  for (var a = arguments.length, b = new Array(a), c = 0; c < a; c++) {b[c] = arguments[c];}
  var d = h(b),
    e = '';
  for (var f in d) {
    if (d[f]) {
      if (typeof d[f] === 'string') {e += e ? ' ' + d[f] : d[f];} else if (typeof d[f] === 'object') {
        var g = d[f];
        for (var i in g) {
          var j = g[i];
          e += e ? ' ' + j : j;
        }
      }
    }
  }
  return e;
}

/** ...................
 * Create an stylex object, this is done by compiled and will caused error
 * if it exits in runtime code.
 *
 * E.g:
 * const styles = stylex.create({
 *     button: {
 *         color: "var(--accent)",
 *         backgroundColor: "var(--secondary-color)",
 *         ...
 *     },
 * })
 *
 * */
stylex.create = function (...args) {
  throw new Error('stylex.create should never be called. It should be compiled away.');
};

/**
 * Override a style property of style object, given by logic condition
 * It's mean: If true => color = "red", otherwise color = "blue".
 *
 * E.g:
 *
 * const {color} = props;
 *
 * return (
 *      <div
 *          className={stylex.dedupe(
 *              {
 *                  position: "relative",
 *                  color: "red",
 *              },
 *              color === "blue" ? {
 *                  color: "red",
 *              } : {},
 *          )}
 *      >
 *          Component
 *      </div>
 * ).
 * */
stylex.dedupe = function () {
  return stylex.apply(undefined, arguments);
};

/**
 * Compose multiple styles object into one.
 * */
stylex.compose = function () {
  for (var a = arguments.length, b = new Array(a), c = 0; c < a; c++) {b[c] = arguments[c];}
  return h(b);
};

/**
 * Create a keyframes animation
 * e.g:
 * const styles = stylex.create({
 *     root: {
 *         position: "relative",
 *         animationName: stylex.keyframes({
 *               '0%': {
 *                   transform: 'translateY(0)'
 *               },
 *               '28%': {
 *                   transform: 'translateY(-5px)'
 *               },
 *               '44%': {
 *                   transform: 'translateY(0)',
 *               },
 *           })
 *     }
 * })
 * => will be transformed to:
 * ...animationName: "sdert25s", <== animation name
 * and an keyframes animation with name "sdert25s".
 * */
stylex.keyframes = function (a) {
  throw new Error('stylex.keyframes should never be called. It should be compiled away.');
};

/**
 * Inject compiled styles to css stylesheet if need (if it's never injected before).
 * */
stylex.inject = function (a, c, d = null) {
  d === void 0 && (d = null), !g && (g = !0), CometStyleXSheet.rootStyleSheet.insert(a, c, d);
};

/**
 * For quick uses.
 * */
stylex.absoluteFill = {
  bottom: 0,
  boxSizing: 'border-box',
  right: 0,
  position: 'absolute',
  left: 0,
  top: 0
};

/**
 * For quick uses.
 * */
stylex.absoluteCenter = {
  boxSizing: 'border-box',
  left: '50%',
  position: 'absolute',
  top: '50%',
  transform: 'translate(-50%, -50%)'
};

/**
 * For quick uses.
 * */
stylex.blockBase = {
  borderStyle: 'solid',
  borderWidth: 0,
  boxSizing: 'border-box',
  display: 'block',
  flexGrow: 1,
  flexShrink: 1,
  margin: 0,
  padding: 0,
  position: 'relative',
  zIndex: 0
};

/**
 * For quick uses.
 * */
stylex.inlineBase = Object.assign({}, stylex.blockBase, {
  display: 'inline'
});

/**
 * For quick uses.
 * */
stylex.buttonBase = {
  appearance: 'none',
  backgroundColor: 'transparent',
  borderStyle: 'solid',
  borderWidth: 0,
  boxSizing: 'border-box',
  margin: 0,
  padding: 0,
  position: 'relative',
  textAlign: 'inherit',
  zIndex: 0
};

/**
 * For quick uses.
 * */
stylex.flexBase = {
  alignItems: 'stretch',
  borderStyle: 'solid',
  borderWidth: 0,
  boxSizing: 'border-box',
  display: 'flex',
  flexDirection: 'column',
  flexGrow: 1,
  flexShrink: 1,
  justifyContent: 'space-between',
  margin: 0,
  minHeight: 0,
  minWidth: 0,
  padding: 0,
  position: 'relative',
  zIndex: 0
};

/**
 * For quick uses.
 * */
stylex.flexInlineBase = Object.assign({}, stylex.flexBase, {
  display: 'inline-flex'
});

/**
 * For quick uses.
 * */
stylex.linkBase = {
  backgroundColor: 'transparent',
  backgroundImage: 'none',
  boxSizing: 'border-box',
  color: 'inherit',
  cursor: 'pointer',
  position: 'relative',
  textDecoration: 'none',
  zIndex: 0
};

/**
 * For quick uses.
 * */
stylex.listBase = {
  boxSizing: 'border-box',
  listStyle: 'none',
  marginBottom: 0,
  marginTop: 0,
  paddingLeft: 0
};

/**
 * For quick uses.
 * */
stylex.visuallyHidden = {
  clip: 'rect(0, 0, 0, 0)',
  clipPath: 'inset(50%)',
  height: 1,
  overflow: 'hidden',
  position: 'absolute',
  width: 1
};

module.exports = stylex;
