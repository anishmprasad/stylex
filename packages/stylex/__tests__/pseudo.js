/* eslint-env jest */
const babelCliTransform = require("./babel-cli-transforms");

it("pseudo", () => {
  const input = `
    import stylex from '@ladifire-opensource/stylex';
    const styles = stylex.create({
      root: {
        color: "red",
        backgroundColor: "blue",

        ":before": {
          color: "blue"
        },
      },
    });
        var a = styles('root', 'button');
  `;
  const {code} = babelCliTransform(input);

  console.log("___code", code);

  expect(code).toMatchSnapshot();
});
