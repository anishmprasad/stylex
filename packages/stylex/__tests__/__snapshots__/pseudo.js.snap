// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`string 1`] = `
"\\"use strict\\";

var _stylex = _interopRequireDefault(require(\\"@ladifire-opensource/stylex\\"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { \\"default\\": obj }; }

var styles = {
  root: {
    color: \\"aaaeyvqo\\",
    backgroundColor: \\"jab5yxxb\\",
    \\"::before\\": {
      color: \\"yaavwbwo\\"
    }
  }
};
var a = styles('root', 'button');"
`;
